scalyr_agent/monitor_utils/k8s.py

```diff
@@ -566,42 +580,209 @@ def process_object( self, obj ):
+    """
+
+    def __init__( self, api_url="https://kubernetes.default", verify_api_queries=True, cache_expiry_secs=30, cache_purge_secs=300, namespaces_to_ignore=None ):
+        """
+        @param api_url: the url for querying the k8s api
+        @param verify_api_queries: whether to verify queries to the k8s api
+        @param cache_expiry_secs: the number of secs to wait before updating the cache
+        @param cache_purge_secs: the number of seconds to wait before purging old controllers from the cache
+        @param namespaces_to_ignore: a list of namespaces to ignore
+        @type api_url: str
+        @type verify_api_queries: bool
+        @type cache_expiry_secs: int or float
+        @type cache_purge_secs: int or float
+        @type namespaces_to_ignore: list[str]
+        """
+        self.api_url = api_url
+        self.verify_api_queries = verify_api_queries
+        self.cache_expiry_secs = cache_expiry_secs
+        self.cache_purge_secs = cache_purge_secs
+        self.namespaces_to_ignore = namespaces_to_ignore
+
+    def __str__( self ):
+        return "\n\tapi_url: %s\n\tverify_api_queries: %s\n\tcache_expiry_secs: %d\n\tcache_purge_secs: %d\n\tnamespaces_to_ignore: %s\n" % ( str(self.api_url),
+            str( self.verify_api_queries), self.cache_expiry_secs, self.cache_purge_secs, str( self.namespaces_to_ignore ) )
+
+    def need_new_k8s_object( self, new_config ):
+        """
+        Determines if a new KubernetesApi object needs to created for the cache based on the new config
+        @param new_config: The new config options
+        @type new_config: _CacheConfig
+        @return: True if a new KubernetesApi object should be created based on the differences between the
+                 current and the new config.  False otherwise.
+        """
+        return self.api_url != new_config.api_url or self.verify_api_queries != new_config.verify_api_queries
+
+    def need_new_filters( self, new_config ):
+        """
+        Determines if new node and namespace filters need to be created based on the new config
+        @param new_config: The new config options
+        @type new_config: _CacheConfig
+        @return: True if new filters need to be created based on the differences between the current and the new config.
+                 False otherwise
+        """
+        return self.namespaces_to_ignore != new_config.namespaces_to_ignore
+
+
 class KubernetesCache( object ):
 
-    def __init__( self, k8s, logger, cache_expiry_secs=30, cache_purge_secs=300, namespaces_to_ignore=None ):
+    def __init__( self, api_url="https://kubernetes.default", verify_api_queries=True, cache_expiry_secs=30, cache_purge_secs=300, namespaces_to_ignore=None, start_caching=True ):
+
+        self._lock = threading.Lock()
+        self._namespace_filter = None
+        self._node_filter = None
+        self._k8s = None
+        self._cache_expiry_secs=cache_expiry_secs
```

Comment:
You should just read these values (`cache_expiry_secs` and `cache_purge_secs`) from `self._config`.  Having them in two places can lead to bugs.

Reply:
I was in two minds about this.  The reason I did it like this was to differentiate "current state" from "configuration options".

The idea being that threads can happily change the configuration options without affecting state and vice versa, and then having a fixed point at which configurations options are turned in to state, allowing a fixed state throughout the caching operations.

Happy to update this, but perhaps moving all cache state (k8s, filters and expiry) to a separate helper class would make the separation I was aiming for clearer?

Reply:
Yeah, I was debating this approach as well.. having a separate class that gets all of the state.  If you think it's a better approach, yes, then please go ahead and do that.